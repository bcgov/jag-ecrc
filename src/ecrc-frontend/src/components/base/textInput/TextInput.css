.label {
  display: flex;
  flex-wrap: wrap;
  align-items: baseline;
}

.textinput_label {
  font-weight: 700;
  font-style: normal;
  font-size: 16px;
  color: #333333;
  text-align: left;
  line-height: 16px;
}

.textinput_editable_gray {
  padding: 1px 0px 1px 0px;
  box-sizing: border-box;
  top: 30px;
  width: 100%;
  height: 40px;
  background-color: #f2f2f2;
  font-weight: 400;
  font-style: normal;
  font-size: 16px;
  text-decoration: none;
  color: #494949;
  text-align: left;
  border-style: inset;
}

.textinput_non_editable_gray {
  top: 30px;
  width: 100%;
  height: 40px;
  background-color: #f2f2f2;
  font-weight: 400;
  font-style: normal;
  font-size: 16px;
  text-decoration: none;
  color: #494949;
  text-align: left;
  border-style: solid;
  border-color: #b1b1b1;
  border-width: 2px;
}

.textinput_editable_white {
  padding: 1px 0px 1px 0px;
  box-sizing: border-box;
  top: 30px;
  width: 100%;
  height: 40px;
  background-color: white;
  font-weight: 400;
  font-style: normal;
  font-size: 16px;
  text-decoration: none;
  color: #494949;
  text-align: left;
  border-style: inset;
}

.mandatory {
  padding: 0px 0px 0px 3px;
  color: red;
  font-weight: 500;
  font-size: 16px;
}

.note {
  font-size: 14px;
  font-weight: 400;
}

.error {
  color: red;
}

input[type="radio"],
input[type="checkbox"] {
  margin-right: 5px;
}

input[type="text"] {
  padding-left: 5px;
}
