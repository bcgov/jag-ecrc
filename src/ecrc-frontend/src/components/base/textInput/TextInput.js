/* eslint-disable react/jsx-one-expression-per-line */
import React from "react";
import PropTypes from "prop-types";
import "./TextInput.css";

export const TextInput = ({
  textInput: {
    label,
    id,
    note,
    textInputStyle,
    value,
    placeholder,
    options,
    isRequired,
    errorMsg
  },
  onChange
}) => {
  let asterisk = "";
  if (isRequired) {
    asterisk = (
      <span id="asterisk" className="mandatory">
        *
      </span>
    );
  }

  let labelPart = null;
  if (label) {
    labelPart = (
      <label htmlFor={id} className="label">
        <div className="textinput_label">{label}</div>
        {asterisk}&nbsp;
        <span className="note">{note}</span>
      </label>
    );
  }

  let textStyle = "textinput_editable_white";
  const validStyles = [
    "textinput_editable_gray",
    "textinput_non_editable_gray",
    "textinput_editable_white"
  ];

  if (validStyles.includes(textInputStyle)) {
    textStyle = textInputStyle;
  }

  if (textStyle === "textinput_non_editable_gray")
    return (
      <div>
        {labelPart && labelPart}
        <input
          className={`${textStyle}`}
          type="text"
          id={id}
          defaultValue={value}
          readOnly
          onChange={onChange}
        />
        <span className="error">{errorMsg}</span>
      </div>
    );

  if (options) {
    return (
      <div>
        {labelPart && labelPart}
        <select className={textStyle} id={id} value={value} onChange={onChange}>
          {options.map(item => {
            return (
              <option key={item.name} value={item.name}>
                {item.name}
              </option>
            );
          })}
        </select>
      </div>
    );
  }

  return (
    <div>
      {labelPart && labelPart}
      <input
        className={`${textStyle}`}
        type="text"
        id={id}
        defaultValue={value}
        placeholder={placeholder}
        onChange={event => onChange(event.target.value)}
      />
      <span className="error">{errorMsg}</span>
    </div>
  );
};

TextInput.propTypes = {
  textInput: PropTypes.shape({
    label: PropTypes.string,
    id: PropTypes.string.isRequired,
    note: PropTypes.string,
    textInputStyle: PropTypes.string,
    value: PropTypes.string,
    placeholder: PropTypes.string,
    options: PropTypes.arrayOf(
      PropTypes.shape({
        name: PropTypes.string.isRequired
      })
    ),
    isRequired: PropTypes.bool,
    errorMsg: PropTypes.string
  }),
  onChange: PropTypes.func
};

TextInput.defaultProps = {
  textInput: {
    placeholder: "",
    note: "",
    errorMsg: ""
  },
  onChange: () => {}
};

export default TextInput;
