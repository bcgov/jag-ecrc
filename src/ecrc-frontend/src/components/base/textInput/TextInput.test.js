/* eslint-disable react/jsx-props-no-spreading */
import React from "react";
import { create } from "react-test-renderer";
import { shallow } from "enzyme";

import { TextInput } from "./TextInput";

describe("TextInput Component", () => {
  const textInput = {
    label: "button label",
    id: "textInputId",
    value: "init value"
  };

  const onChange = jest.fn();

  test("Matches the snapshot", () => {
    const textInputBox = create(
      <TextInput textInput={{ ...textInput }} onChange={onChange} />
    );
    expect(textInputBox.toJSON()).toMatchSnapshot();
  });

  test("Renders the textInput when the style is not editable", () => {
    const textInputBox = shallow(
      <TextInput
        textInput={{
          ...textInput,
          textInputStyle: "textinput_non_editable_gray"
        }}
        onChange={onChange}
      />
    );
    expect(textInputBox.find("div input").prop("readOnly")).toBe(true);
  });

  test("Renders the textInput when options provided", () => {
    const textInputBox = shallow(
      <TextInput
        textInput={{
          ...textInput,
          options: [{ name: "name1" }, { name: "name2" }]
        }}
        onChange={onChange}
      />
    );

    const event = { target: { value: "val" } };
    textInputBox.find("select").simulate("change", event);

    expect(onChange).toHaveBeenCalled();
  });

  test("Renders the textInput correctly when isRequired is set", () => {
    const textInputBox = shallow(
      <TextInput
        textInput={{
          ...textInput,
          isRequired: true
        }}
        onChange={onChange}
      />
    );
    expect(textInputBox.exists("#asterisk")).toEqual(true);
  });
});
