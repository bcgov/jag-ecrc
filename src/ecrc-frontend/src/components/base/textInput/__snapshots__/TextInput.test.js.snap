// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput Component Matches the snapshot 1`] = `
<div>
  <label
    className="label"
    htmlFor="textInputId"
  >
    <div
      className="textinput_label"
    >
      button label
    </div>
    
    Â 
    <span
      className="note"
    />
  </label>
  <input
    className="textinput_editable_white"
    defaultValue="init value"
    id="textInputId"
    onChange={[Function]}
    type="text"
  />
  <br />
  <span
    className="error"
  />
</div>
`;
